// mini-webpack/lib/Complier.js
const fs = require("fs");
const ejs = require("ejs");
const path = require("path");

class Complier {
    constructor(config){
        this.config = config;
        // {路径:code}
        this.modules = {};
    }
    run(){
        this.buildModule();
        // todo
        // this.emitFile();
    }
    buildModule(){
        // webpack入口
        let code = this.getSource(this.config.entry);
        // 存 {路径:code}
        this.modules[this.config.entry] = code;
        // test
        console.log(this.modules);
    }
    // 字符串路径
    getSource(modulePath){
        return fs.readFileSync(modulePath, "utf8");
    }

    // tODO
    emitFile(){
        // 1.读取EJS模板
        let templatePath = path.resolve(__dirname, "main.ejs");
        let template = fs.readFileSync(templatePath, "utf8");
        // console.log(template);
        // 2.利用变量替换模板中的内容
        let resultCode = ejs.render(template, {entryId: this.config.entry, modules: this.modules});
        // console.log(resultCode);
        // 3.将最终的内容写入到文件中
        // 3.1拿到输出的目录
        let outputDir = this.config.output.path;
        // console.log(outputDir);
        if(!fs.existsSync(outputDir)){
            fs.mkdirSync(outputDir);
        }
        let outputPath = path.resolve(outputDir, this.config.output.filename)
;
        fs.writeFileSync(outputPath, resultCode);
    }
}
module.exports = Complier;
